File: RUST/src/a_star.rs
use std::collections::{BTreeMap, BinaryHeap};
use std::cmp::Ordering;


type Position = (usize, usize);

// Node for the priority queue
#[derive(Copy, Clone, Eq, PartialEq)]
struct Node {
    position: Position,
    f_score: usize,  // f_score = g_score + h_score
    g_score: usize,  // g_score = cost from start to current node
}

// Custom ordering for the priority queue (min-heap)
impl Ord for Node {
    fn cmp(&self, other: &Self) -> Ordering {
        // Compare by f_score (min-heap)
        other.f_score.cmp(&self.f_score)
            // Break ties with g_score (prefer nodes closer to start when f_scores are equal)
            .then_with(|| other.g_score.cmp(&self.g_score))
    }
}

impl PartialOrd for Node {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

fn is_wall(grid: &Vec<i32>, row: usize, col: usize, cols: usize) -> bool {
    let index = get_index(row, col, cols);
    grid[index] == 1
}

fn get_index(row: usize, col: usize, cols: usize) -> usize {
    row * cols + col
}

// Heuristic function for A* (Manhattan distance with diagonal movement)
fn heuristic(current: Position, goal: Position) -> usize {
    let dx = if current.1 > goal.1 { current.1 - goal.1 } else { goal.1 - current.1 };
    let dy = if current.0 > goal.0 { current.0 - goal.0 } else { goal.0 - current.0 };

    // Using the diagonal shortcut heuristic (same as in the C++ implementation)
    if dx > dy {
        14 * dy + 10 * (dx - dy)
    } else {
        14 * dx + 10 * (dy - dx)
    }
}

fn get_neighbors(
    grid: &Vec<i32>,
    position: Position,
    rows: usize,
    cols: usize,
) -> Vec<(Position, usize)> {
    let (row, col) = position;
    let mut neighbors = Vec::new();

    // Define possible movements (right, down, left, up)
    let directions = [(0, 1), (1, 0), (0, -1), (-1, 0)];

    for &(dy, dx) in &directions {
        let new_row = row as isize + dy;
        let new_col = col as isize + dx;

        // Check bounds
        if new_row < 0 || new_row >= rows as isize || new_col < 0 || new_col >= cols as isize {
            continue;
        }

        let new_row = new_row as usize;
        let new_col = new_col as usize;

        // Skip walls
        if is_wall(grid, new_row, new_col, cols) {
            continue;
        }

        // Add neighbor with weight 1 (all steps cost the same)
        neighbors.push(((new_row, new_col), 1));
    }

    neighbors
}

fn reconstruct_path(
    came_from: &BTreeMap<Position, Position>,
    start: Position,
    end: Position,
) -> Vec<Position> {
    let mut path = Vec::new();
    let mut current = end;

    while current != start {
        path.push(current);
        current = came_from[&current];
    }

    // Add the start position
    path.push(start);

    // Reverse to get path from start to end
    path.reverse();

    path
}

pub fn astar_grid(
    grid: &mut Vec<i32>,
    start: Position,
    end: Position,
    rows: usize,
    cols: usize,
) -> (Option<Vec<Position>>, Vec<usize>) {
    let mut open_set = BinaryHeap::new();
    let mut closed_set = BTreeMap::new();
    let mut g_scores = BTreeMap::new();
    let mut f_scores = BTreeMap::new();
    let mut came_from = BTreeMap::new();
    let mut visited_order = Vec::new();

    // Initialize start node
    g_scores.insert(start, 0);
    f_scores.insert(start, heuristic(start, end));
    open_set.push(Node {
        position: start,
        f_score: heuristic(start, end),
        g_score: 0,
    });

    // Add start position to visited order
    visited_order.push(get_index(start.0, start.1, cols));

    while let Some(current_node) = open_set.pop() {
        let current = current_node.position;

        // If we've already processed this node, skip it
        if closed_set.contains_key(&current) {
            continue;
        }

        // Add current node to visited order if not already visited
        let current_index = get_index(current.0, current.1, cols);
        if !visited_order.contains(&current_index) {
            visited_order.push(current_index);
        }

        // Mark node as processed
        closed_set.insert(current, true);

        // Check if we've reached the goal
        if current == end {
            // Reconstruct the path
            let path = reconstruct_path(&came_from, start, end);

            // Mark path cells in the grid
            for &(row, col) in &path {
                let index = get_index(row, col, cols);
                if grid[index] != 1 {  // Don't mark walls
                    grid[index] = 3;   // Mark as path
                }
            }

            return (Some(path), visited_order);
        }

        // Process neighbors
        for (neighbor, cost) in get_neighbors(grid, current, rows, cols) {
            // Skip already processed nodes
            if closed_set.contains_key(&neighbor) {
                continue;
            }

            // Calculate tentative g_score
            let tentative_g_score = g_scores[&current] + cost;

            // If we found a better path to neighbor
            if !g_scores.contains_key(&neighbor) || tentative_g_score < g_scores[&neighbor] {
                // Update neighbor's scores
                came_from.insert(neighbor, current);
                g_scores.insert(neighbor, tentative_g_score);
                let h_score = heuristic(neighbor, end);
                f_scores.insert(neighbor, tentative_g_score + h_score);

                // Add neighbor to open set
                open_set.push(Node {
                    position: neighbor,
                    f_score: tentative_g_score + h_score,
                    g_score: tentative_g_score,
                });

                // Mark unvisited cells in the grid
                let index = get_index(neighbor.0, neighbor.1, cols);
                if grid[index] == 0 {
                    grid[index] = 2;  // Mark as visited
                }
            }
        }
    }

    // If we get here, there's no path to the goal
    // Mark all visited cells in the grid
    for &index in &visited_order {
        if grid[index] == 0 {
            grid[index] = 2;  // Mark as visited
        }
    }

    (None, visited_order)
}

// Function to get the indexes of nodes in shortest path
pub fn get_path_indexes(path: &Vec<Position>, cols: usize) -> Vec<usize> {
    path.iter()
        .map(|&(row, col)| get_index(row, col, cols))
        .collect()
}

pub fn find_shortest_path(
    mut grid: Vec<i32>,
    start_row: usize,
    start_col: usize,
    end_row: usize,
    end_col: usize,
    rows: usize,
    cols: usize,
) -> (Vec<i32>, Vec<usize>, Vec<usize>) {
    let start = (start_row, start_col);
    let end = (end_row, end_col);

    let (path_opt, visited_order) = astar_grid(&mut grid, start, end, rows, cols);

    // Get path indexes if a path was found
    let path_indexes = match path_opt {
        Some(path) => get_path_indexes(&path, cols),
        None => Vec::new()
    };

    (grid, visited_order, path_indexes)
}

File: Algorithims/A_Star.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <map>
#include <algorithm>

using namespace std;

using Position = pair<int, int>;

struct Node {
    Position position;
    int f_score;
    int g_score;

    // Reverse comparison for min-heap
    bool operator<(const Node& other) const {
        if (f_score == other.f_score)
            return g_score > other.g_score; // prefer lower g_score on tie
        return f_score > other.f_score;
    }
};

int getIndex(int row, int col, int cols) {
    return row * cols + col;
}

bool isWall(const vector<int>& grid, int row, int col, int cols) {
    int index = getIndex(row, col, cols);
    return grid[index] == 1;
}

// Diagonal-based Manhattan heuristic
int heuristic(Position current, Position goal) {
    int dx = abs(current.second - goal.second);
    int dy = abs(current.first - goal.first);
    return dx > dy ? 14 * dy + 10 * (dx - dy) : 14 * dx + 10 * (dy - dx);
}

vector<pair<Position, int>> getNeighbors(const vector<int>& grid, Position pos, int rows, int cols) {
    int row = pos.first, col = pos.second;
    vector<pair<Position, int>> neighbors;

    vector<pair<int, int>> directions = {
        {0, 1}, {1, 0}, {0, -1}, {-1, 0}
        // Uncomment below for diagonal movement:
        //,{1, 1}, {1, -1}, {-1, 1}, {-1, -1}
    };

    for (auto [dy, dx] : directions) {
        int new_row = row + dy;
        int new_col = col + dx;

        if (new_row < 0 || new_row >= rows || new_col < 0 || new_col >= cols)
            continue;

        if (isWall(grid, new_row, new_col, cols))
            continue;

        int cost = (abs(dy) + abs(dx) == 2) ? 14 : 10; // 14 for diagonal, 10 otherwise
        neighbors.push_back({{new_row, new_col}, cost});
    }

    return neighbors;
}

vector<Position> reconstructPath(map<Position, Position>& came_from, Position start, Position end) {
    vector<Position> path;
    Position current = end;

    while (current != start) {
        path.push_back(current);
        current = came_from[current];
    }
    path.push_back(start);
    reverse(path.begin(), path.end());
    return path;
}

pair<optional<vector<Position>>, vector<int>> astarGrid(
    vector<int>& grid, Position start, Position end, int rows, int cols
) {
    priority_queue<Node> open_set;
    map<Position, bool> closed_set;
    map<Position, int> g_scores;
    map<Position, int> f_scores;
    map<Position, Position> came_from;
    vector<int> visited_order;

    g_scores[start] = 0;
    int h_start = heuristic(start, end);
    f_scores[start] = h_start;

    open_set.push({start, h_start, 0});
    visited_order.push_back(getIndex(start.first, start.second, cols));

    while (!open_set.empty()) {
        Node current_node = open_set.top();
        open_set.pop();
        Position current = current_node.position;

        if (closed_set[current])
            continue;

        int current_index = getIndex(current.first, current.second, cols);
        if (find(visited_order.begin(), visited_order.end(), current_index) == visited_order.end()) {
            visited_order.push_back(current_index);
        }

        closed_set[current] = true;

        if (current == end) {
            vector<Position> path = reconstructPath(came_from, start, end);
            for (auto [r, c] : path) {
                int idx = getIndex(r, c, cols);
                if (grid[idx] != 1)
                    grid[idx] = 3;
            }
            return {path, visited_order};
        }

        for (auto [neighbor, cost] : getNeighbors(grid, current, rows, cols)) {
            if (closed_set[neighbor])
                continue;

            int tentative_g = g_scores[current] + cost;

            if (!g_scores.count(neighbor) || tentative_g < g_scores[neighbor]) {
                came_from[neighbor] = current;
                g_scores[neighbor] = tentative_g;
                int h = heuristic(neighbor, end);
                f_scores[neighbor] = tentative_g + h;

                open_set.push({neighbor, tentative_g + h, tentative_g});

                int idx = getIndex(neighbor.first, neighbor.second, cols);
                if (grid[idx] == 0)
                    grid[idx] = 2;
            }
        }
    }

    for (int idx : visited_order) {
        if (grid[idx] == 0)
            grid[idx] = 2;
    }

    return {nullopt, visited_order};
}

vector<int> getPathIndexes(const vector<Position>& path, int cols) {
    vector<int> indexes;
    for (auto [r, c] : path)
        indexes.push_back(getIndex(r, c, cols));
    return indexes;
}

tuple<vector<int>, vector<int>, vector<int>> findShortestPath(
    vector<int> grid, int start_row, int start_col, int end_row, int end_col, int rows, int cols
) {
    Position start = {start_row, start_col};
    Position end = {end_row, end_col};

    auto [opt_path, visited_order] = astarGrid(grid, start, end, rows, cols);

    vector<int> path_indexes;
    if (opt_path)
        path_indexes = getPathIndexes(opt_path.value(), cols);

    return {grid, visited_order, path_indexes};
}
